parser/impl/parser.cpp

```diff
@@ -7,619 +7,664 @@ namespace parser {
 using lexer::Token;
 using lexer::TokenType;
 
-// ---- @kureduro
-
-static std::array<TokenType, 2> primitives{
+static std::array<TokenType, 3> primitives{
     TokenType::IntegerType,
     TokenType::RealType,
+    TokenType::Boolean,
 };
 
 sPtr<ast::Program> Parser::parseProgram() {
     ast::Program programNode;
-    Token currentToken;
-    while ((currentToken = m_lexer.Peek()).type != TokenType::Eof) {
+    Token currentToken = m_lexer.Peek();
+    programNode.begin = currentToken.pos;
+    while (currentToken.type != TokenType::Eof) {
         switch (currentToken.type) {
         case TokenType::Routine:
-            programNode.routines.push_back(parseRoutine());
-            continue;
+            programNode.routines.push_back(parseRoutineDecl());
+            break;
         case TokenType::Var:
-            programNode.variables.push_back(parseVariable());
-            continue;
+            programNode.variables.push_back(parseVariableDecl());
+            break;
         case TokenType::Type:
-            programNode.types.push_back(parseType());
-            continue;
+            programNode.types.push_back(parseTypeDecl());
+            break;
         case TokenType::NewLine:
-            m_lexer.Next();
-            continue;
+            currentToken = m_lexer.Next();
+            break;
+        case TokenType::Eof:
+            break;
         default:
             m_errors.push_back(Error{
                 .pos = currentToken.pos,
-                .message = "Unexpected token",
+                .message = "Unexpected token.",
             });
-            while (m_lexer.Peek().type != TokenType::NewLine)
+            while (m_lexer.Peek().type != TokenType::NewLine &&
```

Comment:
Shouldn't there be some kind of `advance` call?

Reply:
Good catch